### ISSUE TITLE
Epic 4: Missing User Story to optionally notify secondary contacts

### ISSUE THREAD

**Issue by cfritzsche** [corona-warn-app/cwa-wishlist#24]
## What
Under Epic 4, include a new user story E4.03:

As an app user, if I am exposed to the virus, I want to optionally notify secondary contacts, so that I can warn them before more time passes waiting for my test result.

## Why
The [Imperial College study](https://science.sciencemag.org/content/368/6491/eabb6936) outlining the usefulness of Contact tracing apps and [Christian Drosten in his podcast](https://www.ndr.de/nachrichten/info/27-Coronavirus-Update-Handy-Apps-koennen-eine-Perspektive-bieten,podcastcoronavirus172.html) talked about the importance of having the option to „skip one hop“ and directly notify secondary contacts as soon as the primary contact gets notified about a contact to an infected person. In certain circumstances (test centers overloaded or take too long), this could save critical time. They even went so far to say in exponential spread, this could be enabled by default to help stop overloading health care infrastructure.

Relevant quote: „The simple algorithm can easily be refined to be more informative — for example, quarantining areas if local epidemics become uncontrolled, quarantining whole households, or performing second- or third-degree contact tracing if case numbers are rising, which would result in more people being preemptively quarantined.“

## Where
New User Story under [Epic 4 (Exposure)](https://github.com/corona-warn-app/cwa-documentation/blob/master/scoping_document.md#exposure-contact-with-an-infected-person)

## How
See [comment below](https://github.com/corona-warn-app/cwa-documentation/issues/16#issuecomment-633275541)

**Comment by HeiDasGri** [comment]
The main issue is to speed up the warning process. A typical timing will be like this:

- 0h :Person X is infected but doesn't have symptoms yet. I am sitting next to this person in a train for several hours
- 24h: Person X get symptoms (RKI:  "höchste Infektiosität am Tag vor dem Symptombeginn")
- 48h: Person X is going to be tested
- 54h: Person X get its test result and immediately upload all keys
- 54h - 76h: The keys are aggregated on the server on daily base
- 54h - 98h: my app is warning me on daily base
- 72h - 120h: I am becoming dangerous for other people  (RKI: Interquartilsabstand: 3–5 Tage)

looking at this timing, the warning of the App does not always receive me in time to protected my family  members. On the other hand, the warning is early enough for my family to be quarantined by health authority. Taking a look at the current process, there a several way to speed up everything. At it is now time to optimise processes and algorithm.

**Step1** 

The most Important feature, which do not change the user story at all, is mentioned here:
corona-warn-app/cwa-backlog#2

**Step 2** 

The next thing I propose to introduce is to eliminate the delay caused by testing. Here I thing we need to make some general differentiation before:
- Case A: I get a positive test result (we already handle this case as infected) 
- Case B: I have had a close contact to an infected person and will therefore make a test. I am potentially infected (not handled today by the App)
- Case C: I am paranoid or like to scare other people but I am not infected at all. I am doing all kind of tricks, to issue a fake warning. (This case is already handled today due to missing test result)

The problem here is to handle case "B" without opening Pandora's box for case "C". Here an health authority or the existing hotline needs to make checks and then provide the same pin like I receive for a positive test. Doing so the processes are all the same. I upload my keys and other users might receive a warning, but much earlier as before.

**Step 3**

Within step 2 we have the disadvantage, that after a negative test I am not able to revoke the uploaded keys. Also a warning caused by an potential infected make no differences to a more seriouse warning due to positive test. Thus we need to differ in the whole process chain two kinds of keys. The keys issued from infected peoples and the keys issued from potential infected peoples people. 

When issuing the keys without a positive test result, the app provides me two different ways:
- I can confirm my corona case later on by a posive test result
- I can revoke all emitted keys due to a negative test result

On server side both kind of keys are provided. Keys of potential infected people might be deleted later on or be transferred to the keys of infected people later on.

On client side a new kind of yellow warning might occurs. ""You have had contact to a person which fears to have been infected. Keep calm, avoid contacts, and wait a view days, the status will be updated after this person has received an test result"

**Comment by cfritzsche** [comment]
Thanks @HeiDasGri. I think for your scenario the yellow and green notification scheme implemented by the Austrian app would work. https://futurezone.at/apps/neue-stopp-corona-app-die-wichtigsten-fragen-und-antworten/400953755 Unfortunately I could not yet find details on how they do it, maybe using distinct values in the exposure risk score, as proposed by Google.

Apart from the "I have symptoms" semantic, the yellow warning could also be used for "I got a red warning but don't have a test result yet". Both help to speed up the process.

**Comment by daimpi** [comment]
Just as a heads up: the yellow color is currently proposed for notifying when contacts occurred, but risk status remains green: https://github.com/corona-warn-app/cwa-app-android/issues/899#issuecomment-663475550

**Comment by cfritzsche** [comment]
OK, that will be a challenge UX-wise to show the additional details about encounters below the threshold and on top explain that the encounter was with someone who doesn't have a positive test yet.

**Comment by HeiDasGri** [comment]
For me labelling both cases with yellow colour seams to be ok. It indicates that there currently is no high risk, but that something has happens which should be considered. I am sure someone with colour yellow will read careful, why this happens. I recommend, not to use to many colours. Mapping to traffic light colour is something, everybody can understand.

**Comment by bullipirat** [comment]
For secondary contacts, I would think this could be implemented even in the de-centralized setup. For example, if I receive a red high risk notification from the app, I could share and upload this (automatically or manually) as a sort of "possible positive test result", so that my close contacts get a warning.

**Comment by cfritzsche** [comment]
Finally some movement here: Apprently upload from antigen test will be supported soon: "Der Bund will im Laufe des Aprils zudem die technische Möglichkeit schaffen, auch Schnelltestergebnisse in der Corona-Warn-App anzuzeigen und in einem „zeitnah“ folgenden zweiten Schritt auch eine Warn-Funktion durch Schnelltests freischalten. „Hierbei wird den Gewarnten angezeigt, ob die Warnung durch einen PCR- oder durch einen Schnelltest verursacht wurde.“ Der Bund will überdies allen Anbietern von Testzentren die Möglichkeit geben, sich an das System der Warn-App anzuschließen." https://www.handelsblatt.com/politik/deutschland/kampf-gegen-die-pandemie-bund-ruestet-seine-corona-warn-app-mit-check-in-funktion-auf/27026914.html

**Comment by heinezen** [comment]
@cfritzsche I think https://github.com/corona-warn-app/cwa-wishlist/issues/374 is the correct place for comment. Because this issue is not about antigen tests.

----
Corona-Warn-App Open Source Team

**Comment by heinezen** [comment]
I hid a bunch of comments in this thread as off-topic/outdated as they were arguing about the general mechanism of exchanging exposure keys, not about contacting secondary users.

----
Corona-Warn-App Open Source Team

**Comment by cfritzsche** [comment]
@heinezen appreciate your feedback. To me as the issue reporter, the main point is speeding up the time needed to issue warnings. Back in May 2020, antigen tests were not on the horizon, so the entire [discussion](https://github.com/corona-warn-app/cwa-wishlist/issues/24#issuecomment-633275541) was based on other ways to do it. For me, allowing upload on positive antigen test would go a long way towards speeding up the warning.

As maintainers, do you still see value and consider implementing another mechanism of warning down the chain of secondary contacts without antigen test? Then of course let's focus this issue on that.

**Comment by heinezen** [comment]
> As maintainers, do you still see value and consider implementing another mechanism of warning down the chain of secondary contacts without antigen test? Then of course let's focus this issue on that.

If the issue is not closed, then it still is considered as a possible feature internally (maybe not with super high priority). We will bring it up again in a dev meeting to see if it is still viable. It might even go hand in hand with antigen/rapid tests.

----
Corona-Warn-App Open Source Team

**Comment by cfritzsche** [comment]
Dear community, in light of a planned feature to upload keys from positive rapid test, do you still see an option to warn secondary contacts when one user gets a warning as an important feature?

Example:
User A meets User B on 23.03., User A is pre-symptomatic and infectious.
25.03. User A has a running nose.
27.03. User A starts feeling sick and decides for a test.
User B meets User C on 28.03. (earliest that B can be infectious as far as I know)

If B takes a rapid test and uploads the result on 27th itself, B will be warned and cancel the meeting with C or be better protected. If B doesn't get a rapid test on 27th or waits a little longer or has to wait for the PCR (as today), a chain notification from A via B to C would still be good to warn C before he has more contacts.
On the other hand, the warned user now has to understand not one but three warning categories: Your contact is PCR-positive (status quo), your contact is rapid test positive (planned) and your contact was just warned because his contact was positive. How would that work out with the exposure window calculation and the confusion already about red and green warnings? I am not sure if it would be worth the complexity overhead.

Would be great to hear from others. @di-sc @daimpi @HeiDasGri @Monikae @bullipirat 

Best Regards,
Christoph

**Comment by cfritzsche** [comment]
As #403 is closed, let's pull over some of the recent discussion here:

I wrote:

> This issue does not require a central database. It just involves a second kind of risk calculation, based on uploaded keys with a different marking, e.g. UNTESTED_CONTACT, and a way to upload them. The rest of the framework could work as before, apps of warned users would tell them the contact was not tested yet but his contact was. 
> 
> See more details here: https://github.com/corona-warn-app/cwa-wishlist/issues/24#issuecomment-633275541

@Ein-Tim asked:

> would it be possible for the app/ENF to distinguish between direct contacts and secondary contacts in any form?
> Or rephrased question:
> Can TEKs be marked:
> 
> POSITIVE_TESTED
> and
> SECONDARY_CONTACT
> 
> so that the app which downloads the keys knows wether this is a key from a confirmed positive tested person or only from a secondary contact?
> 
> I hope you know what I mean.

@heinezen answered:

> The app could attach any additional metadata to the uploaded keys which could be stored by the server. So in general: Yes, something like this is technically possible.

**Comment by Ein-Tim** [comment]
@cfritzsche 

Thanks for mirroring and sorry for the duplicate. I didn't yet had the time to read through your issue here tbh. 
But nice to know that it should be possible.

**Comment by di-sc** [comment]
> Example:
> User A meets User B on 23.03., User A is pre-symptomatic and infectious.
> 25.03. User A has a running nose.
> 27.03. User A starts feeling sick and decides for a test.
> User B meets User C on 28.03. (earliest that B can be infectious as far as I know)
> 
> If B takes a rapid test and uploads the result on 27th itself, B will be warned and cancel the meeting with C or be better protected. If B doesn't get a rapid test on 27th or waits a little longer or has to wait for the PCR (as today), a chain notification from A via B to C would still be good to warn C before he has more contacts.
> On the other hand, the warned user now has to understand not one but three warning categories: Your contact is PCR-positive (status quo), your contact is rapid test positive (planned) and your contact was just warned because his contact was positive. How would that work out with the exposure window calculation and the confusion already about red and green warnings? I am not sure if it would be worth the complexity overhead.

The math:
The data from the RKI (https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Steckbrief.html):
Avg. Incubation time: 5-6 days (first symtoms!?) with high variation
Avg. Infectious time: starts 1-2 days before first symtoms with high variation

So on day 3 A can infect B and B may be infectious on day 6. If A has first symtoms on day 5 and gets results early on day 6, we are fine, he will be notified before he can infect others.
That is also the statement of the RKI to that point. 

But:
1. The data above have a high variation. There are also cases where people are infections on day 1 after being infected (see https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Steckbrief.html)
2. The scenario above is an optimistic one, there may be delays everywhere. Are A and B aware that every second counts? Does it realistically describe how people behave?
2a. Is A going to the doctor on observing first symptoms? And if he has symptoms on friday, won't he wait until monday to go to the doctor? ("May be it is only the flu and will go away?")
2b. What if B does ignore his notification at all because he has no symptoms?
2c. A is uploading his positive test results one week later after making the test.
3. The virus mutates with worse infection data

All these cases could be handled by the CWA by notifying the complete contact branch of A and all other delays that might occur can be handled. During the delays the virus may spread but when A registers his positive test result the complete branch can be notified because all contact data is available (and if all people involved have installed the CWA!) I think this is a big advantage of the CWA.
It is my personal opinion.

**Comment by cfritzsche** [comment]
> 2b. What if B does ignore his notification at all because he has no symptoms?
> 
> 2c. A is uploading his positive test results one week later after making the test.

You are right. It would still add value if everything doesn't go ideally. But I think because of the decentral solution, C can only be warned by B. So if A or B don't cooperate, this feature doesn't help.

And how do you judge the complexity for the user?

**Comment by di-sc** [comment]
@cfritzsche 
Is it too complex to the user? This is difficult to say for me. Until now I only saw the green notification, contact to 2 persons that notified their infection but without risk for me. Because I wouldn't get the notifiction very often I don't know what's possible, anyway. So if I will see a yellow notification with a hint that I **may** have had a contact to an infected person and the instructions to make a test, and isolate me until then, I would treat it as okay and follow that instruction.
There won't be different notifications in parallel, right? This would be confusing.
If the instructions, what to do are clear, it would be okay for me.
(A possible confusing thing to some people might be the green notification with the hint to do nothing. I think this has irritated some people)

**Comment by di-sc** [comment]
It might be confusing if C gets first the secondary notification from B makes a test that is negative and some days later, if the test results from B was positive C gets a further notification from B.
This should be avoided.

**Comment by ndegendogo** [comment]
> It might be confusing if C gets first the secondary notification from B makes a test that is negative and some days later, if the test results from B was positive C gets a further notification from B.
This should be avoided.

@di-sc actualky, this **cannot** be avoided.
All downloaded keys are anonymous. cwa does not know if the new warning is from the same person, the same 'chain', or totally unrelated.

**Comment by ndegendogo** [comment]
And if I do a test with negative result my infection state can still change every day.

**Comment by jucktnich** [comment]
@ndegendogo it would be possible to upload the keys with the origin keys. But this would need big technical changes.

**Comment by di-sc** [comment]
> > It might be confusing if C gets first the secondary notification from B makes a test that is negative and some days later, if the test results from B was positive C gets a further notification from B.
> > This should be avoided.
> 
> @di-sc actualky, this **cannot** be avoided.
> All downloaded keys are anonymous. cwa does not know if the new warning is from the same person, the same 'chain', or totally unrelated.

You are right, person C can also receive such notifications from others, he cannot differentiate between them. So it makes no difference only that the notifications may come in short time intervals if B is fast in making a test.
Okay, I think this is acceptable.

**Comment by di-sc** [comment]
[Corona-Warn-App Wikipedia - Delays](https://de.wikipedia.org/wiki/Corona-Warn-App#Gr%C3%BCnde_f%C3%BCr_Verz%C3%B6gerungen) mentions some problems with delays in the past. May be most are solved and/or improved. But the probability that delays occur will increase with and increasing infection rate, I think that is obvious.

**Comment by heinezen** [comment]
@di-sc 

The information in the article is outdated. A few corrections:

> Bis der Befund im Labor vorliegt und die getesteten Personen die Nachricht über ihr Ergebnis nach Hause bekommen, kann es zwei Tage und länger dauern.

The whole testing pipeline (PCR test at doctor -> CWA test result server) has been improved and should take 22-28 hours on average (see https://github.com/corona-warn-app/cwa-website/pull/832#issuecomment-771928934), but no longer than 48 hours. We currently encourage reports for cases where the test result does not appear after two days in the respective [FAQ entry](https://www.coronawarn.app/de/faq/#qr_test).

> [...] die nur ein Mal täglichen Abrufe der Tagesschlüssel durch die Smartphones der Kontakt-Personen [...]

The app now checks the result servers 6 times per day (= roughly every 4 hours).

----
Corona-Warn-App Open Source Team

**Comment by di-sc** [comment]
@heinezen 
Absolutely right and good improvements.
Critical is the time from having symptomes or receiving a notification up to the time sending the own result about a positive test result. So there is the time waiting for a test result which is ~1 day but additionally there would be delays to find an appointment for making a test, etc. With the current speed the infection rate is increasing during the 3rd wave I believe it is rather likely that the delay will increase.

**Comment by di-sc** [comment]
Some numbers (taking the average data for incubation and infectious time from [RKI-Corona](https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Steckbrief.html) for calculation):
<br/>  
![covid-infectious2_delay1](https://user-images.githubusercontent.com/54937900/112796461-4a27fd80-906a-11eb-9019-b90f8fb37f37.jpg)
With a delay of 1 day B receives the notification at the same day he becomes infectious so there is a possibility that C won't receive the notification early enough. But C will send his/her notification early enough to notify D.
<br/> 
![covid-infectious2_delay2](https://user-images.githubusercontent.com/54937900/112796743-b0148500-906a-11eb-9dc0-b8b494885ef8.jpg)
With a delay of 2 days B and C might not send their notifications early enough but for D/E and following contacts it is again fine.
<br/> 

![covid-infectious2_delay3](https://user-images.githubusercontent.com/54937900/112796885-e3571400-906a-11eb-8ff8-79df8e1ca439.jpg)
With a delay of 3 days all notifications wont't receive the contacts early enough and there are 2 additional days that an infection may occur. Additionally the notification is received at the end of incubation time so when notified the contact may have symptoms already, which makes the CoronaWarnApp more or less useless.

**Comment by jucktnich** [comment]
The incubation time is most likely 5-6 days, so it's not a problem to wait 5-6 days since contact for testing (you can get tested early, but your not infectious.

**Comment by di-sc** [comment]
@jucktnich 
> The incubation time is most likely 5-6 days, so it's not a problem to wait 5-6 days since contact for testing (you can get tested early, but your not infectious).

[RKI-Coronavirus-Steckbrief](https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Steckbrief.html;jsessionid=0C897BBBD90F62AA5F59558295AB3760.internet081?nn=13490888#doc13776792bodyText3) states in section 3 that infectious time starts 1-2 days before symptoms (incubation time). (This may have changed with new variants and studies.)

**Comment by ndegendogo** [comment]
@di-sc, @jucktnich do you know if a PCR test is expected to indicate an infection as soon as I am contagious / even before I have any symptoms?

**Comment by di-sc** [comment]
> @di-sc, @jucktnich do you know if a PCR test is expected to indicate an infection as soon as I am contagious / even before I have any symptoms?

The PCR is much more precise, yes. I don't know if it can detect start of infectious phase but it is much better than rapid tests (and I think the diagram I posted above is a proof of that).
[CoronaVirus Update - 84](https://www.ndr.de/nachrichten/info/84-Nicht-auf-Tests-und-Impfungen-verlassen,audio869676.html) (~min 48)

**Comment by Monikae** [comment]
> @di-sc, @jucktnich do you know if a PCR test is expected to indicate an infection as soon as I am contagious / even before I have any symptoms?

Oh yes, absolutely! Even before you are contagious! A typical/average case is like this: 
Get infected 
2 days: not contagious, still PCR-negative, no symptoms 
1 day: PCR already positive, not yet contagious, no symptoms 
2 days: already (highly!) contagious, of course PCR still positive, no symptoms yet 
Now symptoms start, contagious, PCR-positive for around a week. 
After a week of symptoms PCR tests on material taken from the throat are usually negative (even in really sick people), the virus is now mostly in the lungs. 
Now you are either getting better or in hospital. 

That was for the wild type, for the newer variants there seems to be no exact information, yet, but people may be contagious longer. Many countries have made quarantine longer.

**Comment by di-sc** [comment]
Because the problem with rapid antigen tests vs. PCR tests is not directly related to this issue (notify secondary contacts) I created a separate issue for it #536

**Comment by ndegendogo** [comment]
@Monikae Thanks.
Actually, my concern was: how reliable will a PCR test be at this early stage? Especially wrt false negatives?
What I did not want, is: test too early, get a false negative, stop self-isolation, and return to work. Also I was not sure if it is possible / allowed to get a second PCR only two days later (just in case...)
